### PR TITLE
added MultiPolygon GeoJSON constructor function

### PR DESCRIPTION
**Pull request by hkernbach**

### CODE REVIEW

**Comment by hkernbach** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/2405/

**Comment by hkernbach** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/2406/

**Comment by hkernbach** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/2407/

File: arangod/Aql/Functions.cpp [hunk 1]
```diff
@@ -5342,6 +5342,117 @@ AqlValue Functions::GeoPolygon(arangodb::aql::Query* query,
...
+       [[40, 40], [20, 45], [45, 30], [40, 40]]
+    ],
+    [
+        [[20, 35], [10, 30], [10, 10], [30, 5], [45, 20], [20, 35]],
+        [[30, 20], [20, 15], [20, 25], [30, 20]]
+    ]
+  ])
+  */
+
+  if (s.isArray() && s.length() < 2) {
```
Comment (mchacki) [inline arangod/Aql/Functions.cpp]: I think this check is wrong:
It should trigger if `s` is not an Array, or length is smaller then 2:
`!s.isArray() || s.length() < 2`

File: arangod/Aql/Functions.cpp [hunk 2]
```diff
@@ -5342,6 +5342,117 @@ AqlValue Functions::GeoPolygon(arangodb::aql::Query* query,
...
+    ::registerWarning(query, "GEO_MULTIPOLYGON", Result(
+          TRI_ERROR_QUERY_FUNCTION_ARGUMENT_TYPE_MISMATCH,
+          "a MultiPolygon needs at least two Polygons inside."));
+    return AqlValue(arangodb::velocypack::Slice::nullSlice());
+  }
+
+  for (auto const& arrayOfPolygons : VPackArrayIterator(s)) {
+  b.openArray();
```
Comment (mchacki) [inline arangod/Aql/Functions.cpp]: indention

File: arangod/Aql/Functions.cpp [hunk 3]
```diff
@@ -5342,6 +5342,117 @@ AqlValue Functions::GeoPolygon(arangodb::aql::Query* query,
...
+              TRI_ERROR_QUERY_FUNCTION_ARGUMENT_TYPE_MISMATCH,
+              "not an array containing positions"));
+        return AqlValue(arangodb::velocypack::Slice::nullSlice());
+      }
+    }
+    b.close();
+  }
+
+  b.close();
+  b.close();
```
Comment (mchacki) [inline arangod/Aql/Functions.cpp]: please add comment `// base object`

File: arangod/Aql/Functions.cpp [hunk 4]
```diff
@@ -5342,6 +5342,117 @@ AqlValue Functions::GeoPolygon(arangodb::aql::Query* query,
...
+        ::registerWarning(query, "GEO_MULTIPOLYGON", Result(
+              TRI_ERROR_QUERY_FUNCTION_ARGUMENT_TYPE_MISMATCH,
+              "not an array containing positions"));
+        return AqlValue(arangodb::velocypack::Slice::nullSlice());
+      }
+    }
+    b.close();
+  }
+
+  b.close();
```
Comment (mchacki) [inline arangod/Aql/Functions.cpp]: please add comment `// coordinates`

File: arangod/Aql/Functions.cpp [hunk 5]
```diff
@@ -5342,6 +5342,117 @@ AqlValue Functions::GeoPolygon(arangodb::aql::Query* query,
...
+              ::registerWarning(query, "GEO_MULTIPOLYGON", Result(
+                    TRI_ERROR_QUERY_FUNCTION_ARGUMENT_TYPE_MISMATCH,
+                    "a Position needs at least two numeric values"));
+              return AqlValue(arangodb::velocypack::Slice::nullSlice());
+            } else {
+              b.openArray();
+              for (auto const& innercord : VPackArrayIterator(coord)) {
+                if (innercord.isNumber()) {
+                  b.add(VPackValue(innercord.getNumber<double>()));
+                } else if (innercord.isArray()) {
```
Comment (mchacki) [inline arangod/Aql/Functions.cpp]: This fails if the innercord array is empty and does not have a 0 element.

File: arangod/Aql/Functions.cpp [hunk 6]
```diff
@@ -5342,6 +5342,117 @@ AqlValue Functions::GeoPolygon(arangodb::aql::Query* query,
...
+                  if (innercord.at(0).isNumber() && innercord.at(1).isNumber()) {
+                    b.openArray();
```
Comment (mchacki) [inline arangod/Aql/Functions.cpp]: oh and no second element as well.

File: Documentation/Books/AQL/Functions/Geo.md [hunk 7]
```diff
@@ -277,6 +277,32 @@ RETURN GEO_POLYGON([
 @END_EXAMPLE_AQL
 @endDocuBlock aqlGeoPolygon_2
 
+### GEO_MULTIPOLYGON()
+
+`GEO_MULTIPOLYGON(polygons) → geoJson`
+
+Construct a GeoJSON MultiPolygon. Needs at least two polygons in an array (You
```
Comment (Simran-B) [inline Documentation/Books/AQL/Functions/Geo.md]:
```suggestion
Construct a GeoJSON MultiPolygon. Needs at least two Polygons in an array.
```

File: Documentation/Books/AQL/Functions/Geo.md [hunk 8]
```diff
@@ -277,6 +277,32 @@ RETURN GEO_POLYGON([
...
+can find the rules for the Polygon construction above).
```
Comment (Simran-B) [inline Documentation/Books/AQL/Functions/Geo.md]:
```suggestion
See [GEO_POLYGON()](#geopolygon) for the rules of Polygon construction.
```

File: Documentation/Books/AQL/Functions/Geo.md [hunk 9]
```diff
@@ -277,6 +277,32 @@ RETURN GEO_POLYGON([
...
+
+- **polygons** (array): array of polygons
+- returns **geoJson** (object|null): a valid GeoJSON Polygon
```
Comment (Simran-B) [inline Documentation/Books/AQL/Functions/Geo.md]:
```suggestion
- returns **geoJson** (object|null): a valid GeoJSON MultiPolygon
```

File: Documentation/Books/AQL/Functions/Geo.md [hunk 10]
```diff
@@ -277,6 +277,32 @@ RETURN GEO_POLYGON([
...
+
+MultiPolygon:
```
Comment (Simran-B) [inline Documentation/Books/AQL/Functions/Geo.md]:
```suggestion
MultiPolygon comprised of a simple Polygon and a Polygon with hole:
```

**Comment by hkernbach** [comment]
@Simran-B thanks simran!

**Comment by hkernbach** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/2412/

**Pull request merged**